src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBarService/VsInfoBarService.cs

```diff
@@ -0,0 +1,132 @@
+        _vsShell = vsShell;
+        _vsInfoBarFactory = vsInfoBarFactory;
+    }
+
+    public async Task ShowInfoBarAsync(string message, ImageMoniker image, CancellationToken cancellationToken, params InfoBarUI[] items)
+    {
+        Requires.NotNullOrEmpty(message, nameof(message));
+        Requires.NotNull(items, nameof(items));
+
+        if (!_vsShellServices.IsInServerMode)
+        {
+            await _threadingService.SwitchToUIThread(cancellationToken);
+
+            IVsInfoBarHost? host = FindMainWindowInfoBarHost();
+            if (host == null)
+            {
+                return;
+            }
+
+            // We want to avoid posting the same message over and over again, so we remove any existing info bar
+            // with the same message, and add the new one which will cause it to float to the bottom of the host.
+            RemoveInfoBar(message);
+            AddInfoBar(host, message, image, items);
+        }
+    }
+
+    private IVsInfoBarHost? FindMainWindowInfoBarHost()
+    {
+        if (_vsShell.Value is null)
+        {
+            return null;
+        }
+
+        if (ErrorHandler.Failed(_vsShell.Value.GetProperty((int)__VSSPROPID7.VSSPROPID_MainWindowInfoBarHost, out object mainWindowInfoBarHost)))
+        {
+            return null;
+        }
+
+        return mainWindowInfoBarHost as IVsInfoBarHost;
+    }
+
+    private IVsInfoBarUIElement? CreateInfoBarUIElement(string message, ImageMoniker image, InfoBarUI[] items)
+    {
+        if (_vsInfoBarFactory.Value is null)
+        {
+            return null;
+        }
+
+        var actionItems = new List<IVsInfoBarActionItem>(items.Length);
+
+        foreach (InfoBarUI item in items)
+        {
+            switch (item.Kind)
+            {
+                default:
+                case InfoBarUIKind.Button:
+                    Assumes.True(item.Kind == InfoBarUIKind.Button);
+                    actionItems.Add(new InfoBarButton(item.Title));
+                    break;
+                case InfoBarUIKind.Hyperlink:
+                    actionItems.Add(new InfoBarHyperlink(item.Title));
+                    break;
+            }
+        }
+
+        var infoBarModel = new InfoBarModel(
+            message,
+            actionItems.ToArray(),
+            image,
+            isCloseButtonVisible: true);
+
+        return _vsInfoBarFactory.Value.CreateInfoBar(infoBarModel);
+    }
+
+    private void AddInfoBar(IVsInfoBarHost host, string message, ImageMoniker image, InfoBarUI[] items)
+    {
+        IVsInfoBarUIElement? element = CreateInfoBarUIElement(message, image, items);
+        if (element != null)
+        {
+            var entry = new InfoBarEntry(message, element, items, OnClosed);
+            entries.Add(entry);
+            host.AddInfoBar(element);
+        }
+    }
+
+    private void RemoveInfoBar(string message)
+    {
+        // Assumption is that message is "good" enough to uniquely identify problems
```

Comment:
> // Assumption is that message is "good" enough to uniquely identify problems [](http://example.com/codeflow?start=7&length=77)

Are we sure this is the case? Seems like a bug waiting to happen

Reply:
We just want to search if there is an existing InfoBar with the same message (not taking into account buttons/links). We can change this implementation if that is really needed.